core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,108 +69,169 @@ class LogManager(val logDirs: Array[File],
   private val logs = new Pool[TopicPartition, Log]()
   private val logsToBeDeleted = new LinkedBlockingQueue[Log]()
 
-  createAndValidateLogDirs(logDirs)
-  private val dirLocks = lockLogDirs(logDirs)
-  private val recoveryPointCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
-  private val logStartOffsetCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+  val liveLogDirs: ArrayBuffer[File] = createAndValidateLogDirs(logDirs)
+  @volatile private var recoveryPointCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
+  @volatile private var logStartOffsetCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+
+  private def offlineLogDirs = logDirs.filterNot(liveLogDirs.toSet)
+
   loadLogs()
```

Comment:
Previously I think it is OK not to handle failure in `loadLogs()` because the subsequent requests (e.g. FetchRequest, ProducerRequest, LeaderAndIsrRequest) can trigger `handleLogDirFailure(...)`. Handling log directory failure in `loadLogs()` could allow broker to fail faster if all log directories are offline at the cost of slightly more code. Anyway, I have updated the code to handle IO failure in log segments loading.